sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/ScriptTransformation.scala

```diff
@@ -122,27 +91,21 @@ case class ScriptTransformation(
           if (!hasNext) {
             throw new NoSuchElementException
           }
-
-          if (outputSerde == null) {
-            val prevLine = curLine
-            curLine = reader.readLine()
-            if (!ioschema.schemaLess) {
-              new GenericInternalRow(CatalystTypeConverters.convertToCatalyst(
-                prevLine.split(ioschema.outputRowFormatMap("TOK_TABLEROWFORMATFIELD")))
-                .asInstanceOf[Array[Any]])
+          val mutableRow = new SpecificMutableRow(output.map(_.dataType))
+          val raw = outputSerde.deserialize(scriptOutputWritable)
+          scriptOutputWritable = null
+          val dataList = outputSoi.getStructFieldsDataAsList(raw)
+          val fieldList = outputSoi.getAllStructFieldRefs()
+          var i = 0
+          dataList.foreach(element => {
+            if (element == null) {
+              mutableRow.setNullAt(i)
             } else {
-              new GenericInternalRow(CatalystTypeConverters.convertToCatalyst(
-                prevLine.split(ioschema.outputRowFormatMap("TOK_TABLEROWFORMATFIELD"), 2))
-                .asInstanceOf[Array[Any]])
-            }
-          } else {
-            val ret = deserialize()
-            if (!eof) {
-              cacheRow = null
-              cacheRow = deserialize()
+              mutableRow(i) = unwrap(element, fieldList(i).getFieldObjectInspector)
```

Comment:
`unwrap` actually support `StructObjectInspector`, we don't need to extract every field here.
But, I prefer to reuse the `mutableRow`, which mean we don't need to create the `mutableRow` for every call of `next()`.